data_structures/src/superblock.rs

```diff
@@ -592,31 +594,20 @@ where
     let each = v.len() / size;
 
     let mut step_index = 0_usize;
-    let mut step = rand_distribution[step_index] as usize;
+    let mut step = rand_distribution[step_index] as usize % each;
     let mut a = first;
     let mut b = (a + step) % v.len();
-    while hs_subset.len() < size {
-        if !hs_subset.contains(&v[b]) {
-            hs_subset.insert(v[b].clone());
-
-            step_index = (step_index + 1) % rand_distribution.len();
-            step = rand_distribution[step_index] as usize;
-            a = (a + each) % v.len();
-            b = (a + step) % v.len();
-        } else {
-            b = (b + 1) % v.len();
-        }
-    }
+    let mut v_subset = vec![];
+    for _ in 0..size {
+        v_subset.push(v[b].clone());
 
-    hs_subset
-}
+        step_index = (step_index + 1) % rand_distribution.len();
+        step = rand_distribution[step_index] as usize % each;
+        a = (a + each) % v.len();
+        b = (a + step) % v.len();
```

Comment:
`a` and `b` also look redundant to me, as `b` can be probably derived as `((first + i * each) % v.len() + step) % v.len()`  before pushing into `v_subset`.

This is only of a coding style suggestion, both approaches do look right to me. It's only that I personally find the procedural version harder to read than the parametric—which comes with the nice feature of not requiring mutable variables.

Reply:
Thinking about this made me realize that there's a mismatch between the code and my description in the WIP:
- Code: `((first + i * each) % v.len() + step) % v.len()`
- WIP: `(first + i * each + step) % v.len()`

Reply:
Any advantage of the first over the second? My intuition is that both will work, but the second is simpler.

Reply:
> Thinking about this made me realize that there's a mismatch between the code and my description in the WIP:

I think it doesn't matter because `(a % N + b) % N == (a + b) % N`

Reply:
@aesedepece I don't want to break the current implementation while trying to simplify it, I think we should rewrite the whole function from scratch to prevent mistakes. So could you provide some pseudocode for the parametric version?

Reply:
Implemented the version without iterators because I guess it will be easier to port to solidity.